### PR TITLE
finished chapter 4

### PR DESCRIPTION
**Pull request by keuhdall**
### Solutions for Chapter 4

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -471,11 +485,12 @@ Applicatives can be found in many applications:
...
 instance Applicative (Secret e) where
-    pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+  pure :: a -> Secret e a
+  pure = Reward
 
-    (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+  (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
+  (<*>) (Reward f) x = fmap f x
```
Comment (chshersh) [inline src/Chapter4.hs:492]: Nice work on reusing `Functor` instance for `Secret`! :world_map:

File: src/Chapter4.hs [hunk 2]
```diff
@@ -490,6 +505,20 @@ Implement the 'Applicative' instance for our 'List' type.
 -}
 
 
+(<+>) :: List a -> List a -> List a
+(<+>) Empty ys = ys
+(<+>) (Cons x xs) ys = Cons x (xs <+> ys)
```
Comment (chshersh) [inline src/Chapter4.hs:510]: Nice operator you have here :phone:

**Review by chshersh — APPROVED** [review]
Rock-solid implementations for all tasks!
And cograts with finishing all four chapters of our course :clap: :tada: 

We would appreciate your feedback on our course. Let us know, how did you like it :slightly_smiling_face: There's a feedback form at the end of our README, and we would really love to hear back :hugs:

**Comment by keuhdall** [comment]
Just filled the feedback form ! Thank you so much for this initiative, I had a great time doing these exercises and your feedback was very valuable !
Keep up the good work !!!

**Pull request merged**